src/impl_/pyclass/static_type_object.rs

```diff
@@ -24,31 +28,61 @@ pub struct LazyStaticType {
     tp_dict_filled: GILOnceCell<PyResult<()>>,
 }
 
-impl LazyStaticType {
-    /// Creates an uninitialized `LazyStaticType`.
+impl<T: PyClass> StaticTypeObject<T> {
+    /// Creates an uninitialized `StaticTypeObject`.
     pub const fn new() -> Self {
-        LazyStaticType {
-            value: GILOnceCell::new(),
-            initializing_threads: const_mutex(Vec::new()),
-            tp_dict_filled: GILOnceCell::new(),
-        }
+        StaticTypeObject(
+            StaticTypeObjectInner {
+                value: GILOnceCell::new(),
+                initializing_threads: const_mutex(Vec::new()),
+                tp_dict_filled: GILOnceCell::new(),
+            },
+            PhantomData,
+        )
+    }
+
+    /// Gets the type object contained by this `StaticTypeObject`, initializing it if needed.
+    pub fn get_or_init(&self, py: Python<'_>) -> *mut ffi::PyTypeObject {
+        self.get_or_try_init(py).unwrap_or_else(|err| {
+            err.print(py);
+            panic!("failed to create type object for {}", T::NAME)
+        })
     }
 
-    /// Gets the type object contained by this `LazyStaticType`, initializing it if needed.
-    pub fn get_or_init<T: PyClass>(&self, py: Python<'_>) -> *mut ffi::PyTypeObject {
-        fn inner<T: PyClass>() -> *mut ffi::PyTypeObject {
+    /// Gets the type object contained by this `StaticTypeObject`, initializing it if needed.
+    pub(crate) fn get_or_try_init(&self, py: Python<'_>) -> PyResult<*mut ffi::PyTypeObject> {
+        fn inner<T: PyClass>() -> PyResult<*mut ffi::PyTypeObject> {
             // Safety: `py` is held by the caller of `get_or_init`.
             let py = unsafe { Python::assume_gil_acquired() };
             create_type_object::<T>(py)
         }
 
-        // Uses explicit GILOnceCell::get_or_init::<fn() -> *mut ffi::PyTypeObject> monomorphization
```

Comment:
I think this comment is still valid and helpful to understand why `inner` is used instead of a closure.